src/main/java/com/yahoo/bullet/querying/Querier.java

```diff
@@ -0,0 +1,622 @@
+        }
+        return new RateLimitError(rateLimit.getCurrentRate(), config);
+    }
+
+    /**
+     * Returns if this query has a time based window.
+     *
+     * @return A boolean that is true if this is a time based query window.
+     */
+    public boolean isTimeBasedWindow() {
+        Window window = runningQuery.getQuery().getWindow();
+        // No window means duration drives the query => it is time based.
+        return window == null || window.isTimeBased();
+    }
+
+    /**
+     * Terminate the query and return the final result.
+     *
+     * @return The final non-null {@link Clip} representing the final result.
+     */
+    public Clip finish() {
+        Clip result = getResult();
+        Meta meta = new Meta();
+        addMetadata(Concept.QUERY_FINISH_TIME, (k) -> meta.add(k, System.currentTimeMillis()));
+        result.add(meta);
+        return result;
+    }
+
+    @Override
+    public String toString() {
+        return runningQuery.toString();
+    }
+
+    // ********************************* Private helpers *********************************
+
+    private boolean filter(BulletRecord record) {
+        List<Clause> filters = runningQuery.getQuery().getFilters();
+        // Add the record if we have no filters
+        if (filters == null) {
+            return true;
+        }
+        // Otherwise short circuit evaluate till the first filter fails. Filters are ANDed.
+        return filters.stream().allMatch(c -> FilterOperations.perform(record, c));
+    }
+
+    private BulletRecord project(BulletRecord record) {
+        Projection projection = runningQuery.getQuery().getProjection();
+        BulletRecord projected = projection != null ? ProjectionOperations.project(record, projection) : record;
+        return addAdditionalFields(projected);
+    }
+
+    private BulletRecord addAdditionalFields(BulletRecord record) {
+        if (timestampKey != null) {
+            record.setLong(timestampKey, System.currentTimeMillis());
+        }
+        return record;
+    }
+
+    private Meta getResultMetadata() {
+        if (metaKeys.isEmpty()) {
+            return null;
+        }
+        Meta meta = new Meta();
+        addMetadata(Concept.QUERY_ID, (k) -> meta.add(k, runningQuery.getId()));
+        addMetadata(Concept.QUERY_BODY, (k) -> meta.add(k, runningQuery.getQuery().toString()));
+        addMetadata(Concept.QUERY_RECEIVE_TIME, (k) -> meta.add(k, runningQuery.getStartTime()));
+        addMetadata(Concept.RESULT_EMIT_TIME, (k) -> meta.add(k, System.currentTimeMillis()));
+        return meta;
+    }
+
+    private void addMetadata(Concept concept, Consumer<String> action) {
+        Meta.consumeRegisteredConcept(concept, metaKeys, action);
+    }
+
+    private boolean timeIsUp() {
+        // Never add to query.getDuration since it can be infinite (Long.MAX_VALUE)
+        return System.currentTimeMillis() - runningQuery.getStartTime() >= runningQuery.getQuery().getDuration();
```

Comment:
Could you extract this function to RunningQuery class, and change here as runningQuery.timeIsUp()? Thus we can mock it when writing tests.

Reply:
Good idea. Will do